File: tools/coach-runtime-migrate.swift
import Foundation
import AppKit
import Security
import LocalAuthentication

enum RuntimeMigrationOperation: String { case stage = "--stage", release = "--release", hold = "--hold"
    var items: [RuntimeMigrationCredential] {
        switch self {
        case .stage: return RuntimeMigrationCredential.allCases
        case .release: return [.clientGate, .wafToken]
        case .hold: return [.wafToken]
        }
    }
}
enum RuntimeMigrationCredential: String, CaseIterable {
    case openAI, clientGate, wafToken, appPrefix, appID, keyID, issuerID, privateKey
    var legacy: CoachCredential? {
        switch self { case .openAI: return .openAI; case .clientGate: return .clientGate
        case .wafToken: return .wafToken; default: return nil }
    }
    var apple: CoachRuntimeCredential? {
        switch self { case .appPrefix: return .appPrefix; case .appID: return .appID; case .keyID: return .keyID
        case .issuerID: return .issuerID; case .privateKey: return .privateKey; default: return nil }
    }
    func valid(_ bytes: Data) -> Bool { legacy?.accepts(bytes) ?? apple?.valid(bytes) ?? false }
}
enum RuntimeMigrationFailure: Error { case retrieval, format, coordinator }
protocol RuntimeMigrationReader { func read(_ item: RuntimeMigrationCredential) throws -> Data }
protocol RuntimeMigrationCoordinator { func run(_ credentials: [RuntimeMigrationCredential: Data]) throws -> String }

// Read only explicitly enumerated existing items. No write/search/rotation or output of values.
struct NativeRuntimeMigrationReader: RuntimeMigrationReader {
    func read(_ item: RuntimeMigrationCredential) throws -> Data {
        if let legacy = item.legacy { return try NativeCoachCredentialReader().read(legacy) }
        guard let apple = item.apple else { throw RuntimeMigrationFailure.retrieval }
        let context = LAContext(); context.interactionNotAllowed = false
        let query: [String: Any] = [kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: "com.reptoday.coach.production", kSecAttrAccount as String: apple.rawValue,
            kSecAttrSynchronizable as String: false, kSecMatchLimit as String: kSecMatchLimitOne,
            kSecReturnData as String: true, kSecUseAuthenticationContext as String: context]
        var value: CFTypeRef?
        guard SecItemCopyMatching(query as CFDictionary, &value) == errSecSuccess,
              let bytes = value as? Data else { throw RuntimeMigrationFailure.retrieval }
        return bytes
    }
}
private final class RuntimePresentedRead: @unchecked Sendable {
    let reader: any RuntimeMigrationReader, item: RuntimeMigrationCredential
    private let lock = NSLock(); private var value: Result<Data, Error>?
    init(reader: any RuntimeMigrationReader, item: RuntimeMigrationCredential) { self.reader = reader; self.item = item }
    func execute() { let result = Result { try reader.read(item) }; lock.lock(); value = result; lock.unlock() }
    func take() throws -> Data { lock.lock(); let result = value; value = nil; lock.unlock()
        guard let result else { throw RuntimeMigrationFailure.retrieval }; return try result.get() }
}
struct PresentedRuntimeMigrationReader: RuntimeMigrationReader {
    let reader: any RuntimeMigrationReader
    func read(_ item: RuntimeMigrationCredential) throws -> Data {
        guard Thread.isMainThread else { throw RuntimeMigrationFailure.retrieval }
        return try MainActor.assumeIsolated {
            let app = NSApplication.shared; app.setActivationPolicy(.accessory)
            let panel = NSAlert(); panel.messageText = "Rep Today migration Keychain access"
            panel.informativeText = "Reading an approved existing server configuration item. Authorize the macOS prompt for this dedicated helper if it appears. No value is displayed or re-entered."
            panel.addButton(withTitle: "Cancel"); app.activate(ignoringOtherApps: true)
            let work = RuntimePresentedRead(reader: reader, item: item)
            DispatchQueue.main.async { DispatchQueue.global(qos: .userInitiated).async {
                work.execute(); DispatchQueue.main.async { NSApplication.shared.abortModal() }
            } }
            let response = panel.runModal(); panel.window.orderOut(nil)
            guard response == .abort else { throw RuntimeMigrationFailure.retrieval }; return try work.take()
        }
    }
}
func runRuntimeMigration(reader: any RuntimeMigrationReader, coordinator: any RuntimeMigrationCoordinator,
                         operation: RuntimeMigrationOperation) throws -> String {
    var credentials: [RuntimeMigrationCredential: Data] = [:]
    defer { for item in RuntimeMigrationCredential.allCases {
        if var bytes = credentials.removeValue(forKey: item) { bytes.resetBytes(in: 0..<bytes.count) }
    } }
    for item in operation.items {
        var bytes = try reader.read(item)
        guard item.valid(bytes) else { bytes.resetBytes(in: 0..<bytes.count); throw RuntimeMigrationFailure.format }
        credentials[item] = bytes
    }
    return try coordinator.run(credentials)
}
struct RuntimeNodeCoordinator: RuntimeMigrationCoordinator {
    let repository: URL, node: URL
    let operation: RuntimeMigrationOperation
    static let confirmed = "confirmed: approved existing account, zone, Worker and attached origin"
    static let protected = "protected: hostname held closed; existing boundary and rate protections verified"
    static let staged = "staged: reviewed runtime source, SQLite namespace and server bindings verified; hostname held closed"
    static let released = "released: reptoday-variety-language-proxy https://coach.reptoday.app/coach; genuine Apple and live model QA pending"
    static let held = "held: hostname closed; Worker, server bindings and security records preserved"
    static let failures = Set(["input", "auth", "account", "zone", "target", "scope", "rules", "route", "settings",
        "secret", "wrangler", "http", "gate", "rate-plan", "namespace", "revision", "rehold", "unexpected"])
    var expected: [String] {
        if operation == .hold { return [Self.confirmed, Self.held] }
        return [Self.confirmed, Self.protected, operation == .stage ? Self.staged : Self.released]
    }
    // Exact complete transcript, including newline/order: partial, duplicate and arbitrary output fail.
    func sanitized(_ reply: Data, status: Int32) throws -> String {
        guard reply.count <= 8192, let text = String(data: reply, encoding: .utf8), text.hasSuffix("\n") else {
            throw RuntimeMigrationFailure.coordinator
        }
        let lines = String(text.dropLast()).components(separatedBy: "\n")
        if status == 0 { guard lines == expected else { throw RuntimeMigrationFailure.coordinator }; return lines.joined(separator: "\n") }
        guard let last = lines.last, last.hasPrefix("blocked: "), Self.failures.contains(String(last.dropFirst(9))),
              Array(lines.dropLast()) == Array(expected.prefix(lines.count - 1)), lines.count <= expected.count else {
            throw RuntimeMigrationFailure.coordinator
        }
        return "blocked: dedicated runtime migration stopped (\(last.dropFirst(9))); preserve the hold and inspect prerequisites without values"
    }
    func run(_ credentials: [RuntimeMigrationCredential: Data]) throws -> String {
        guard Set(credentials.keys) == Set(operation.items) else { throw RuntimeMigrationFailure.coordinator }
        let process = Process(); process.executableURL = node
        process.arguments = [repository.appendingPathComponent("tools/coach-runtime-migrate.mjs").path, operation.rawValue]
        process.currentDirectoryURL = repository
        var environment = ProcessInfo.processInfo.environment
        for name in ["NODE_OPTIONS", "NODE_DEBUG", "NODE_DEBUG_NATIVE", "OPENAI_API_KEY", "CLIENT_SHARED_SECRET",
                     "ANTHROPIC_API_KEY", "APP_STORE_PRIVATE_KEY"] { environment.removeValue(forKey: name) }
        process.environment = environment
        let input = Pipe(), output = Pipe(); process.standardInput = input; process.standardOutput = output
        process.standardError = FileHandle.nullDevice
        var packet = try JSONSerialization.data(withJSONObject: Dictionary(uniqueKeysWithValues: credentials.map {
            ($0.key.rawValue, String(decoding: $0.value, as: UTF8.self))
        }))
        defer { packet.resetBytes(in: 0..<packet.count) }
        guard packet.count <= 12_288 else { throw RuntimeMigrationFailure.coordinator }
        try process.run()
        let timer = DispatchWorkItem { if process.isRunning { process.terminate() } }
        DispatchQueue.global().asyncAfter(deadline: .now() + 700, execute: timer)
        defer { timer.cancel() }
        do { try input.fileHandleForWriting.write(contentsOf: packet); try input.fileHandleForWriting.close() }
        catch { if process.isRunning { process.terminate() }; throw RuntimeMigrationFailure.coordinator }
        var reply = Data()
        do { while let part = try output.fileHandleForReading.read(upToCount: 4096), !part.isEmpty {
            guard reply.count + part.count <= 8192 else {
                if process.isRunning { process.terminate() }; throw RuntimeMigrationFailure.coordinator
            }; reply.append(part)
        } } catch { if process.isRunning { process.terminate() }; throw RuntimeMigrationFailure.coordinator }
        process.waitUntilExit(); return try sanitized(reply, status: process.terminationStatus)
    }
}
#if !COACH_RUNTIME_MIGRATION_TESTS
@main struct CoachRuntimeMigrationMain {
    @MainActor static func main() {
        let args = Array(CommandLine.arguments.dropFirst())
        guard args.count == 3, let operation = RuntimeMigrationOperation(rawValue: args[0]) else {
            print("usage: launch tools/migrate-coach-runtime.sh with --stage, --release or --hold; never pass credentials"); exit(64)
        }
        do {
            let result = try runRuntimeMigration(reader: PresentedRuntimeMigrationReader(reader: NativeRuntimeMigrationReader()),
                coordinator: RuntimeNodeCoordinator(repository: URL(fileURLWithPath: args[1], isDirectory: true),
                    node: URL(fileURLWithPath: args[2]), operation: operation), operation: operation)
            print(result); if result.hasPrefix("blocked:") { exit(78) }
        } catch {
            print("blocked: dedicated runtime migration stopped before a verified result; preserve the hold; no credential output"); exit(78)
        }
    }
}
#endif

File: tools/coach-runtime-migrate.mjs
// Dedicated native Keychain child. Preparation tests inject every external boundary.
import fs from 'node:fs/promises';
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import { createPrivateKey } from 'node:crypto';
import { execFileSync } from 'node:child_process';
import { Cloudflare, DeploymentFailure, TARGET, CUSTOM, RATE, HOLD,
  checkSettings, checkDomains, checkRoutes, ensureRule, verifyProtection, entrypoint,
  verifyClosedWorker, credentialsFromPacket, inspectionCredentialsFromPacket,
  readWranglerOAuth, stageWithWrangler, gateProbes } from './coach-production-deploy.mjs';

const requireThat = (condition, code) => { if (!condition) throw new DeploymentFailure(code); };
const identifier = value => typeof value === 'string' && /^[a-f0-9]{32}$/.test(value);
const revisionValid = value => typeof value === 'string' && /^[a-f0-9]{40}$/.test(value);
const CLASS = 'CoachAuthenticationState', TAG = 'coach-security-v1';
const MODE = 'app-attest-storekit-v1';
const APPLE = Object.freeze({ appPrefix: 'APP_ATTEST_APP_PREFIX', appID: 'APP_STORE_APP_ID',
  keyID: 'APP_STORE_KEY_ID', issuerID: 'APP_STORE_ISSUER_ID', privateKey: 'APP_STORE_PRIVATE_KEY' });
export const runtimeSecretNames = Object.freeze(['OPENAI_API_KEY', 'CLIENT_SHARED_SECRET', ...Object.values(APPLE)]);
export const runtimeFailures = Object.freeze(['input', 'auth', 'account', 'zone', 'target', 'scope',
  'rules', 'route', 'settings', 'secret', 'wrangler', 'http', 'gate', 'rate-plan', 'namespace', 'revision', 'rehold', 'unexpected']);
export const runtimeLines = Object.freeze({
  confirmed: 'confirmed: approved existing account, zone, Worker and attached origin',
  protected: 'protected: hostname held closed; existing boundary and rate protections verified',
  staged: 'staged: reviewed runtime source, SQLite namespace and server bindings verified; hostname held closed',
  released: `released: ${TARGET.worker} ${TARGET.origin}; genuine Apple and live model QA pending`,
  held: 'held: hostname closed; Worker, server bindings and security records preserved',
});

export function runtimePacket(packet, operation) {
  requireThat(['--stage', '--release', '--hold'].includes(operation), 'input');
  if (operation === '--hold') return inspectionCredentialsFromPacket(packet);
  if (operation === '--release') {
    requireThat(packet && Object.keys(packet).sort().join(',') === 'clientGate,wafToken', 'input');
    inspectionCredentialsFromPacket({ wafToken: packet.wafToken });
    requireThat(typeof packet.clientGate === 'string' && /^[a-f0-9]{64}$/.test(packet.clientGate), 'input');
    return packet;
  }
  requireThat(packet && Object.keys(packet).sort().join(',') ===
    'appID,appPrefix,clientGate,issuerID,keyID,openAI,privateKey,wafToken', 'input');
  credentialsFromPacket({ openAI: packet.openAI, clientGate: packet.clientGate, wafToken: packet.wafToken });
  requireThat(/^[A-Z0-9]{10}$/.test(packet.appPrefix ?? '') && /^[A-Z0-9]{10}$/.test(packet.keyID ?? '') &&
    typeof packet.appID === 'string' && /^[1-9][0-9]{0,15}$/.test(packet.appID) && Number.isSafeInteger(Number(packet.appID)) &&
    typeof packet.issuerID === 'string' && /^[a-f0-9]{8}-(?:[a-f0-9]{4}-){3}[a-f0-9]{12}$/i.test(packet.issuerID) &&
    typeof packet.privateKey === 'string' && packet.privateKey.length <= 4096 &&
    packet.privateKey.startsWith('-----BEGIN PRIVATE KEY-----\n'), 'input');
  try {
    const key = createPrivateKey(packet.privateKey);
    requireThat(key.asymmetricKeyType === 'ec' && key.asymmetricKeyDetails.namedCurve === 'prime256v1', 'input');
  } catch { throw new DeploymentFailure('input'); }
  return packet;
}

export class RuntimeCloudflare extends Cloudflare {
  async accountRequest(endpoint, method = 'GET', body) {
    const worker = `/accounts/${this.account}/workers/scripts/${TARGET.worker}`;
    if (method === 'GET' && identifier(this.account) &&
        endpoint === `/accounts/${this.account}/workers/durable_objects/namespaces`) {
      return this.request(this.oauth, endpoint);
    }
    if (method === 'PUT' && endpoint === worker + '/secrets') {
      requireThat(identifier(this.account) && body?.type === 'secret_text' && runtimeSecretNames.includes(body.name) &&
        Object.keys(body).sort().join(',') === 'name,text,type' && typeof body.text === 'string' && body.text.length <= 4096, 'secret');
      return this.request(this.oauth, endpoint, method, body);
    }
    // The migration never attaches a hostname, edits DNS or creates another Worker.
    requireThat(method === 'GET' && body === undefined, 'scope');
    return super.accountRequest(endpoint, method, body);
  }
}

export function runtimeConfig(repository, revision, diagnostics = false) {
  requireThat(revisionValid(revision), 'revision');
  return { name: TARGET.worker, main: path.join(repository, 'proxy/src/coach-auth-worker.js'),
    compatibility_date: '2026-01-01', compatibility_flags: ['nodejs_compat'], workers_dev: false,
    preview_urls: false, routes: [], logpush: false, observability: { enabled: false }, send_metrics: false,
    vars: { COACH_AUTH_MODE: MODE, COACH_AUTH_SOURCE_REV: revision, ...(diagnostics ? { COACH_AUTH_GUARD_DIAGNOSTICS: '1' } : {}) },
    alias: { 'node-fetch': path.join(repository, 'proxy/src/apple-fetch.js') },
    durable_objects: { bindings: [{ name: 'COACH_AUTH_STATE', class_name: CLASS }] },
    migrations: [{ tag: TAG, new_sqlite_classes: [CLASS] }] };
}

export function checkRuntimeSettings(settings, revision, complete = false, diagnostics = false) {
  requireThat(settings && Array.isArray(settings.bindings), 'settings');
  const seen = new Set(); let state;
  for (const binding of settings.bindings) {
    requireThat(binding && typeof binding.name === 'string' && !seen.has(binding.name), 'settings'); seen.add(binding.name);
    if (binding.type === 'secret_text') requireThat(runtimeSecretNames.includes(binding.name), 'settings');
    else if (binding.name === 'COACH_AUTH_MODE') requireThat(binding.type === 'plain_text' && binding.text === MODE, 'settings');
    else if (binding.name === 'COACH_AUTH_GUARD_DIAGNOSTICS')
      requireThat(binding.type === 'plain_text' && binding.text === '1', 'settings');
    else if (binding.name === 'COACH_AUTH_SOURCE_REV') requireThat(binding.type === 'plain_text' &&
      revisionValid(binding.text) && (!revision || binding.text === revision), 'revision');
    else if (binding.name === 'COACH_AUTH_STATE') {
      requireThat(binding.type === 'durable_object_namespace' && binding.class_name === CLASS &&
        identifier(binding.namespace_id) && (!binding.script_name || binding.script_name === TARGET.worker) &&
        (!binding.environment || binding.environment === 'production'), 'namespace'); state = binding;
    } else throw new DeploymentFailure('settings');
  }
  requireThat(state && seen.has('COACH_AUTH_MODE') && seen.has('COACH_AUTH_SOURCE_REV'), 'settings');
  // Pre-stage inspection permits either approved state; a pinned release must match the explicit option.
  if (revision) requireThat(seen.has('COACH_AUTH_GUARD_DIAGNOSTICS') === diagnostics, 'settings');
  requireThat((settings.observability === undefined || settings.observability?.enabled === false) &&
    (settings.logpush === undefined || settings.logpush === false) &&
    (settings.tail_consumers === undefined || Array.isArray(settings.tail_consumers) && !settings.tail_consumers.length), 'settings');
  if (settings.observability) for (const section of ['logs', 'traces']) {
    const value = settings.observability[section];
    requireThat(value === undefined || value && value.enabled === false && value.persist !== true &&
      value.invocation_logs !== true && (!value.destinations || Array.isArray(value.destinations) && !value.destinations.length), 'settings');
  }
  if (complete) requireThat(runtimeSecretNames.every(name => seen.has(name)), 'secret');
  return state.namespace_id;
}

function checkSecrets(secrets, complete = false) {
  requireThat(Array.isArray(secrets) && secrets.every(secret => secret?.type === 'secret_text' && runtimeSecretNames.includes(secret.name)) &&
    new Set(secrets.map(secret => secret.name)).size === secrets.length, 'secret');
  requireThat(['OPENAI_API_KEY', 'CLIENT_SHARED_SECRET'].every(name => secrets.some(secret => secret.name === name)), 'secret');
  if (complete) requireThat(runtimeSecretNames.every(name => secrets.some(secret => secret.name === name)), 'secret');
}

async function confirm(cf) {
  const accounts = await cf.accountRequest('/accounts');
  requireThat(Array.isArray(accounts) && accounts.length === 1 && identifier(accounts[0].id), 'account'); cf.account = accounts[0].id;
  const zones = await cf.accountRequest(`/zones?name=${TARGET.zone}&account.id=${cf.account}&status=active`);
  requireThat(Array.isArray(zones) && zones.length === 1 && zones[0].name === TARGET.zone && zones[0].status === 'active' &&
    zones[0].account?.id === cf.account && identifier(zones[0].id), 'zone');
  requireThat(/^Free(?:\b|\s)/i.test(zones[0].plan?.name ?? ''), 'rate-plan'); cf.setScope(cf.account, zones[0].id);
  const scripts = await cf.accountRequest(`/accounts/${cf.account}/workers/scripts`);
  requireThat(Array.isArray(scripts) && scripts.filter(script => script.id === TARGET.worker).length === 1, 'target');
  checkDomains(await cf.accountRequest(`/accounts/${cf.account}/workers/domains`), cf.account, cf.zone, true);
  checkRoutes(await cf.accountRequest(`/zones/${cf.zone}/workers/routes`));
  // Existing protections must already be present. Do not create capacity or fill missing rules.
  const custom = await entrypoint(cf, CUSTOM);
  const hold = custom?.rules.find(rule => rule.ref === HOLD.ref);
  requireThat(hold && typeof hold.enabled === 'boolean', 'rules'); await verifyProtection(cf, hold.enabled);
  return `/accounts/${cf.account}/workers/scripts/${TARGET.worker}`;
}

async function verifyRuntime(cf, worker, revision, complete, diagnostics = false) {
  const namespace = checkRuntimeSettings(await cf.accountRequest(worker + '/settings'), revision, complete, diagnostics);
  await verifyClosedWorker(cf, worker);
  const scripts = await cf.accountRequest(`/accounts/${cf.account}/workers/scripts`);
  requireThat(Array.isArray(scripts) && scripts.filter(script => script.id === TARGET.worker && script.migration_tag === TAG).length === 1, 'namespace');
  // Cloudflare's official namespace API exposes class/script/use_sqlite. Missing fields stop.
  const namespaces = await cf.accountRequest(`/accounts/${cf.account}/workers/durable_objects/namespaces`);
  requireThat(Array.isArray(namespaces) && namespaces.filter(item => item.id === namespace).length === 1, 'namespace');
  const owned = namespaces.find(item => item.id === namespace);
  requireThat(owned.class === CLASS && owned.script === TARGET.worker && owned.use_sqlite === true &&
    namespaces.filter(item => item.script === TARGET.worker && item.class === CLASS).length === 1, 'namespace');
  checkDomains(await cf.accountRequest(`/accounts/${cf.account}/workers/domains`), cf.account, cf.zone, true);
  checkRoutes(await cf.accountRequest(`/zones/${cf.zone}/workers/routes`));
  return namespace;
}

export async function migrateRuntime({ cf, credentials, operation, revision, stageWorker, probe, report = () => {}, diagnostics = false }) {
  runtimePacket(credentials, operation);
  requireThat(operation === '--hold' || revisionValid(revision), 'revision');
  const worker = await confirm(cf); report(runtimeLines.confirmed);
  let previousNamespace;
  if (operation !== '--hold') {
    const settings = await cf.accountRequest(worker + '/settings');
    if (operation === '--stage' && !settings.bindings?.some(binding => binding.name === 'COACH_AUTH_STATE')) checkSettings(settings, true);
    else previousNamespace = await verifyRuntime(cf, worker, operation === '--release' ? revision : null, operation === '--release', diagnostics);
    checkSecrets(await cf.accountRequest(worker + '/secrets'), operation === '--release');
    await verifyClosedWorker(cf, worker);
  }
  await ensureRule(cf, CUSTOM, HOLD); await verifyProtection(cf, true);
  if (operation === '--hold') { report(runtimeLines.held); return; }
  report(runtimeLines.protected);
  if (operation === '--stage') {
    requireThat(typeof stageWorker === 'function', 'wrangler'); await stageWorker(cf.account);
    const namespace = await verifyRuntime(cf, worker, revision, false, diagnostics);
    requireThat(!previousNamespace || namespace === previousNamespace, 'namespace');
    const configured = await cf.accountRequest(worker + '/secrets'); checkSecrets(configured);
    for (const [source, name] of Object.entries(APPLE)) if (!configured.some(secret => secret.name === name)) {
      await cf.accountRequest(worker + '/secrets', 'PUT', { name, type: 'secret_text', text: credentials[source] });
    }
    checkSecrets(await cf.accountRequest(worker + '/secrets'), true);
    requireThat(await verifyRuntime(cf, worker, revision, true, diagnostics) === namespace, 'namespace');
    await verifyProtection(cf, true); report(runtimeLines.staged); return;
  }
  const custom = await entrypoint(cf, CUSTOM), hold = custom.rules.find(rule => rule.ref === HOLD.ref);
  try {
    requireThat(typeof probe === 'function', 'gate');
    await cf.zoneRequest(`/rulesets/${custom.id}/rules/${hold.id}`, 'PATCH', { ...HOLD, enabled: false });
    await verifyProtection(cf, false); await probe(credentials.clientGate);
  } catch (error) {
    try { await ensureRule(cf, CUSTOM, HOLD); await verifyProtection(cf, true); }
    catch { throw new DeploymentFailure('rehold'); }
    throw error;
  }
  report(runtimeLines.released);
}

export async function runtimeGateProbes(gate, fetchImpl = fetch) {
  await gateProbes(gate, fetchImpl);
  // A fixed forged proof cannot reach enrollment, Apple or the provider. No real prompt/proof.
  try {
    const response = await fetchImpl(TARGET.origin, { method: 'POST', redirect: 'error', signal: AbortSignal.timeout(5000),
      headers: { 'Content-Type': 'application/json', 'X-RepToday-Coach-Auth': '{}' }, body: '{}' });
    requireThat(response.status === 401 && response.redirected === false && response.body, 'gate');
    const reader = response.body.getReader(); let bytes = Buffer.alloc(0);
    try { while (true) { const part = await reader.read(); if (part.done) break;
      requireThat(bytes.length + part.value.length <= 256, 'gate'); bytes = Buffer.concat([bytes, part.value]); } }
    finally { await reader.cancel().catch(() => {}); }
    const result = JSON.parse(bytes.toString('utf8'));
    requireThat(Object.keys(result).join(',') === 'error' && result.error === 'unauthorized', 'gate');
  } catch { throw new DeploymentFailure('gate'); }
}

export function runtimeArguments(args) {
  requireThat(args.length >= 1 && args.length <= 2 && ['--stage', '--release', '--hold'].includes(args[0]), 'input');
  const diagnostics = args.length === 2;
  requireThat(!diagnostics || args[1] === '--auth-guard-diagnostics' && args[0] !== '--hold', 'input');
  return { operation: args[0], diagnostics };
}

async function main() {
  requireThat(!process.stdin.isTTY, 'input');
  const { operation, diagnostics } = runtimeArguments(process.argv.slice(2));
  const repository = path.dirname(path.dirname(fileURLToPath(import.meta.url)));
  let bytes = Buffer.alloc(0), packet;
  try {
    for await (const chunk of process.stdin) { requireThat(bytes.length + chunk.length <= 12_288, 'input');
      const next = Buffer.concat([bytes, chunk]); bytes.fill(0); bytes = next; }
    packet = runtimePacket(JSON.parse(bytes.toString('utf8')), operation);
  } catch { throw new DeploymentFailure('input'); } finally { bytes.fill(0); }
  let revision;
  try { revision = execFileSync('git', ['rev-parse', 'HEAD'], { cwd: repository, encoding: 'utf8', stdio: ['ignore', 'pipe', 'ignore'] }).trim(); }
  catch { throw new DeploymentFailure('revision'); }
  const oauth = await readWranglerOAuth();
  await migrateRuntime({ cf: new RuntimeCloudflare(oauth, packet.wafToken), credentials: packet, operation, revision, diagnostics,
    stageWorker: account => stageWithWrangler(repository, account, oauth, root => runtimeConfig(root, revision, diagnostics)),
    probe: gate => runtimeGateProbes(gate), report: line => process.stdout.write(line + '\n') });
}
if (process.argv[1] && path.resolve(process.argv[1]) === fileURLToPath(import.meta.url)) {
  main().catch(error => { const code = error instanceof DeploymentFailure && runtimeFailures.includes(error.code) ? error.code : 'unexpected';
    process.stdout.write('blocked: ' + code + '\n'); process.exitCode = 78; });
}
